ros_ws/src/autonomous/reinforcement_learning/scripts/training_node.py

```diff
@@ -0,0 +1,160 @@
+
+    def get_episode_summary(self):
+        average_episode_length = sum(
+            self.episode_length_history) / len(self.episode_length_history)
+        average_cumulative_reward = sum(
+            self.cumulative_reward_history) / len(self.cumulative_reward_history)
+
+        result = "Episode " + str(self.episode_count) + ": " \
+            + str(self.episode_length).rjust(3) + " steps (" + str(int(average_episode_length)).rjust(3) + " avg), " \
+            + "return: " + "{0:.1f}".format(self.cumulative_reward).rjust(5) \
+            + " (" + "{0:.1f}".format(average_cumulative_reward).rjust(5) + " avg)"
+
+        episode_duration_real = time.time() - self.episode_start_time_real
+        episode_duration_sim = rospy.Time.now().to_sec() - self.episode_start_time_sim
+        if episode_duration_real != 0:
+            result += ", time: {0:.1f}x".format(episode_duration_sim / episode_duration_real)  # nopep8
+        if episode_duration_sim != 0:
+            result += ", laser: {0:.1f} Hz".format(float(self.episode_length) / episode_duration_sim)  # nopep8
+        return result
+
+    def on_complete_episode(self):
+        self.episode_length_history.append(self.episode_length)
+        self.cumulative_reward_history.append(self.cumulative_reward)
+        self.episode_result_publisher.publish(
+            reward=self.cumulative_reward, length=int(
+                self.episode_length))
+
+        rospy.loginfo(self.get_episode_summary())
+        self.episode_start_time_real = time.time()
+        self.episode_start_time_sim = rospy.Time.now().to_sec()
+
+        self.episode_count += 1
+        self.episode_length = 0
+        self.cumulative_reward = 0
+
+        if self.episode_count % 50 == 0:
+            self.policy.save()
+            rospy.loginfo("Model parameters saved.")
+
+    def on_receive_laser_scan(self, message):
+        new_state = self.convert_laser_message_to_tensor(message)
+
+        if self.state is not None:
+            self.check_car_orientation()
+            reward = self.get_reward()
+            self.cumulative_reward += reward
+            self.on_complete_step(self.state, self.action, reward, new_state)
+
+        if self.is_terminal_step or self.episode_length >= self.max_episode_length:
+            self.drive_forward = random.random() > 0.5
+            reset_car.reset_random(
+                max_angle=math.pi / 180 * 20,
+                max_offset_from_center=0.2,
+                forward=self.drive_forward)
+            self.is_terminal_step = False
+            self.state = None
+            if self.episode_length != 0:
+                self.on_complete_episode()
+        else:
+            self.state = new_state
+            self.action = self.select_action(new_state)
+            self.perform_action(self.action)
+            self.episode_length += 1
+            self.total_step_count += 1
+
+    def on_complete_step(self, state, action, reward, next_state):
+        pass
```

Comment:
Thats a pretty nice method :D

Reply:
It's like an abstract method, not sure if this is the correct way to do it. It would also be possible to raise an exception ("not implemented").